### DISCUSSION TITLE
PointCloudLayer appears to sub-sample

### DISCUSSION THREAD

**Discussion by nmonahelis** [visgl/deck.gl#3914] [Q&A]
I'm attempting to load in and render pointcloud data to accompany the xviz stream in streetscape.gl. After discovering deck.gl had support for rendering pointclouds using the PointCloudLayer class, I was able to easily import the classes and get a PointCloudLayer rendered. 

However, I noticed that the data appears to be sub-sampled with some sort of occupancy grid, as there seems to be uniform spacing between the points. In order to confirm that all of the positions were being loaded into memory with deck.gl, I added a counter in the getPosition() method and made sure the number of positions matched the total number of vertices (based on the LAZ file). I also loaded the .laz file into CloudCompare to ensure the data is not compromised. All of the points seem to be rendered in CloudCompare. 

Since I was testing this in a streetscape.gl project, I wanted to see if I could reproduce this in a pure deck.gl example. 

I cloned the PointCloudLayer example from the website (https://deck.gl/#/examples/core-layers/point-cloud-layer) and loaded in one of our .laz files and I'm still getting the same results. 

Here's what it looks like using the PointCloudLayer example:

<img width="1440" alt="Screen Shot 2019-11-21 at 12 11 47 PM" src="https://user-images.githubusercontent.com/40645193/69374946-0e9e4780-0c5c-11ea-843b-74858da481ad.png">

Here's what it should look like rendered using CloudCompare:

<img width="1066" alt="Screen Shot 2019-11-21 at 12 20 54 PM" src="https://user-images.githubusercontent.com/40645193/69375077-4dcc9880-0c5c-11ea-86f1-65f209b38685.png">

In case you wanted to know how I'm loading in the .laz files, I am using the LASLoader class from loaders.gl - the loader seems to be working just fine as I'm able to see the object with the proper data and it doesn't appear to be missing any of it. The COORDINATE_SYSTEM is LNGLAT. 

I've tried initializing PointCloudLayer in two ways: Passing the LAZ object by itself as well as creating an array of the positions, colors and normals like so:  

`{ "position" : ..., "color": ..., "normal": ... }`

Both have the same results. 

Anyone know what could be causing this issue?

**Comment by nmonahelis** [comment]
After doing some digging, I've discovered that this was caused by floating point precision. `parse-las.js` is currently using the `Float32Array` typed array for the `positions` data. I changed it to use `Float64Array` and I'm able to render the full pointcloud now in streetscape: 

![image](https://user-images.githubusercontent.com/40645193/69389380-b4fc4400-0c80-11ea-92f6-f43620c81ec4.png)

**Comment by Pessimistress** [comment]
This should be fixed in loaders.gl (or at least make it configurable). @xintongxia

**Comment by yuanzh0u** [comment]
Hi, @nmonahelis. I have met the same problem. The position values of the las points lost precision during the loading.
![image](https://user-images.githubusercontent.com/39179041/75863478-82dfae00-5e3b-11ea-8979-6d709087e5f3.png)

I've tried to change the `Float32Array` to `Float64Array` in `@loaders.gl/las/src/lib/parse-las.js` like this:
```javascript
......
 parseLASChunked(arraybuffer, skip, (decoder, header) => {
    if (!originalHeader) {
      originalHeader = header;
      const total = header.totalToRead;
      positions = new Float64Array(total * 3);
      // positions = new Float32Array(total * 3);
......
```
But problem didn't go away, instead the console promote an error like so:
```
assert.js?54c0:3 Uncaught (in promise) Error: index out of picking color range
    at assert (assert.js?54c0:3)
    at PointCloudLayer.calculateInstancePickingColors (layer.js?2bb8:580)
    at Attribute.updateBuffer (attribute.js?2db2:192)
    at AttributeManager._updateAttribute (attribute-manager.js?98e5:313)
    at AttributeManager.update (attribute-manager.js?98e5:140)
    at PointCloudLayer._updateAttributes (layer.js?2bb8:523)
    at PointCloudLayer._updateState (layer.js?2bb8:758)
    at PointCloudLayer._initialize (layer.js?2bb8:695)
    at LayerManager._initializeLayer (layer-manager.js?acc3:417)
    at LayerManager._updateSublayersRecursively (layer-manager.js?acc3:355)
```
And the attributes of point cloud data doesn't seem to be changed at all:
```
data: Object
attributes:
COLOR_0: {value: Uint8Array(3232168), size: 4}
POSITION: {value: Float32Array(2424126), size: 3}
classification: {value: Uint8Array(808042), size: 1}
instanceColors: {value: Uint8Array(3232168), size: 4}
instanceNormals: Float32Array(2424126) [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, …]
instancePositions: {value: Float32Array(2424126), size: 3}
intensity: {value: Uint16Array(808042), size: 1}
```
Would you please elaborate on how you over come this problem?

**Comment by Pessimistress** [comment]
As of `@loaders.gl/las@2.3.8` you can explicitly load the point cloud in [64-bit positions](https://loaders.gl/modules/las/docs/api-reference/las-loader#options):


```js
registerLoaders(LASWorkerLoader);

new PointCloudLayer({
  // ...
  data: LAZ_FILE,
  loadOptions: {
    las: {fp64: true}
  }
})
```
